dbms/src/Operators/AutoPassThroughHashAggContext.cpp

```diff
@@ -0,0 +1,182 @@
+        trySwitchFromAdjustState(agg_process_info->block.rows(), agg_process_info->hit_row_cnt);
+        break;
+    }
+    case State::PreHashAgg:
+    {
+        aggregator->executeOnBlock(*agg_process_info, *many_data[0], 0);
+        trySwitchBackAdjustState(agg_process_info->block.rows());
+        break;
+    }
+    case State::PassThrough:
+    {
+        const auto total_rows = agg_process_info->block.rows();
+        onBlockForceStreaming(agg_process_info->block);
+        trySwitchBackAdjustState(total_rows);
+        break;
+    }
+    case State::Selective:
+    {
+        aggregator->executeOnBlockOnlyLookup(*agg_process_info, *many_data[0], 0);
+        auto pass_through_rows = agg_process_info->not_found_rows;
+        const auto total_rows = agg_process_info->block.rows();
+        if (!pass_through_rows.empty())
+        {
+            RUNTIME_CHECK(!agg_process_info->block.info.selective);
+            auto new_block = getPassThroughBlock(agg_process_info->block);
+            new_block.info.selective = std::make_shared<std::vector<UInt64>>(std::move(pass_through_rows));
+            pushPassThroughBuffer(new_block);
+        }
+        trySwitchBackAdjustState(total_rows);
+        break;
+    }
+    default:
+    {
+        __builtin_unreachable();
+    }
+    };
+    RUNTIME_CHECK(agg_process_info->allBlockDataHandled());
+}
+
+void AutoPassThroughHashAggContext::onBlockForceStreaming(Block & block)
+{
+    auto new_block = getPassThroughBlock(block);
+    pushPassThroughBuffer(new_block);
+}
+
+void AutoPassThroughHashAggContext::forceState()
+{
+    if (many_data[0]->need_spill)
+        state = State::PassThrough;
+}
+
+Block AutoPassThroughHashAggContext::getData()
+{
+    if unlikely (!already_start_to_get_data)
+    {
+        // No need to handle situation when agg keys_size is zero.
+        // Shouldn't use auto pass through hashagg in that case.
+        already_start_to_get_data = true;
+        RUNTIME_CHECK(!merging_buckets);
+        merging_buckets = aggregator->mergeAndConvertToBlocks(many_data, /*final=*/true, /*max_threads=*/1);
+    }
+
+    // merging_buckets still can be nullptr when HashMap is empty.
+    if (merging_buckets)
+    {
+        return merging_buckets->getData(/*concurrency_index=*/0);
+    }
+    return {};
+}
+
+void AutoPassThroughHashAggContext::trySwitchFromInitState()
+{
+    if (many_data[0]->bytesCount() > INIT_STATE_HASHMAP_THRESHOLD)
+        state = State::Adjust;
+}
+
+void AutoPassThroughHashAggContext::trySwitchFromAdjustState(size_t total_rows, size_t hit_rows)
+{
+    adjust_processed_rows += total_rows;
+    adjust_hit_rows += hit_rows;
+
+    LOG_DEBUG(
+        log,
+        "adjust state info: processed: {}, hit: {}, limit: {}",
+        adjust_processed_rows,
+        adjust_hit_rows,
+        adjust_row_limit);
+
+    if (adjust_processed_rows < adjust_row_limit)
+        return;
+
+    float hit_rate = static_cast<double>(adjust_hit_rows) / adjust_processed_rows;
```

Comment:
Maybe we can all use double type instead?

Reply:
done